### ISSUE TITLE
Preloading modules with magic comments does not result in modulepreload injection

### ISSUE THREAD

**Issue by sinemacula-ben** [webpack/webpack#16838]
# Bug report

Webpack magic comments do not inject the correct HTML syntax for preloading modules.

**What is the current behavior?**

Preloading a module results in the following HTML injection:
```
<link rel="preload" as="script" href="http://example.com/js/123.456789.chunk.mjs">
```

**If the current behavior is a bug, please provide the steps to reproduce.**
I believe this is a bug, but I am aware that it could be to do with my configuration...

My configuration is large so I will only include the bit that I think is relevant:

```
target: [ 'web', 'es6' ],
output: {
    filename: 'js/[id].[contenthash].mjs',
    chunkFilename: 'js/[id].[contenthash].chunk.mjs'
}
```

I also enabled `experiments.outputModule` and added `module: true` to the `output`, but I got errors with this, so reverted.

And in one of my files, I have the following line:
```
component: () => import(/* webpackChunkName: "about", webpackPreload: true */ '../views/About.vue')
```

**What is the expected behavior?**
I would expect the injected HTML to look like this instead:

```
<link rel="modulepreload" href="http://example.com/js/123.456789.chunk.js">
```

I may be mistaken here as obviously `/* webpackPrefetch: true */` would not work with a module, so perhaps this is intentional, and module preloading is not meant to be supported. If this is the case, then I apologise - I did search the documentation and Google from top to bottom before posting here...

**Other relevant information:**
webpack version:  5.76.1
Node.js version: 19.7
Operating System: WSL 2 (Linux on Windows)
Additional tools:

**Comment by sinemacula-ben** [comment]
Hi @alexander-akait - I am using the [HTML Webpack Plugin](https://github.com/jantimon/html-webpack-plugin), however, I handle the injection manually using an HTML `.ejs` template as opposed to using `inject: true`.

Upon further inspection, I think I have misunderstood the documentation here. From what I now understand, the magic comments will provide Webpack context for how to inject the the `script` tags into the DOM upon evaluation, NOT compilation. The `scripts` inject upon compilation are the 'entry/initial' scripts, and then when the JS is evaluated, it will load any additional modules needed as and when required. For example, navigating to a new route, the JS will inject a new script tag into the DOM to load the module(s) necessary to run the route. The purpose of the magic comments is to advise Webpack whether to preload/prefetch these modules at entry point based on the likelihood they will be needed. Given modules cannot be 'prefetched', they can only be preloaded using `modulepreload`, the magic comments do not really apply to them...

So, with that in mind, I think this is more of a feature request than a bug report. This is how I would expect it to work:

**See the following:**

When a compilation is set to `target: ['es6']`, we handle the magic comments slightly differently:

This:
```
import(/* webpackPreload: true */ 'ChartingLibrary');
```

Will yield and injection of the following into the DOM:
```
<link rel="modulepreload" href="http://example.com/js/123.456789.chunk.js">
```

However, this:
```
import(/* webpackPrefetch: true */ 'ChartingLibrary');
```

Will either be ignored, or will yield the same as the above (I can't decide) - this will make this functionality backward compatible if people upgrade to ES6.

**Comment by alexander-akait** [comment]
But we don't generate HTML here, all this needs to be handled and generated in `html-webpack-plugin`, `prefetch` works as expected (you can find `<link rel="prefetch" as="script" href="/dist/src_module1_js.bundle.js">`), but `webpackPrefetch` doesn't equal `<link rel="modulepreload" href="http://example.com/js/123.456789.chunk.js">`

> Prefetch/preload doesn’t work in the entrypoint for me. What’s the problem?
> The webpack runtime only takes care of prefetch/preload for on-demand-loaded chunks. When prefetch/preload is used in import()s in a entry chunk the html generation is resposible for adding the <link> tags to the HTML. The stats json data provides information in entrypoints[].childAssets.

Ref: https://medium.com/webpack/link-rel-prefetch-preload-in-webpack-51a52358f84c

**Comment by alexander-akait** [comment]
Oh, I see a bug, when you have:
```
experiments: {
  outputModule: true
}
```

We generate:
```
<link type="module" charset="utf-8" rel="preload" as="script" href="/dist/src_module2_js.bundle.js">
```

But we need generate:
```
<link rel="modulepreload" href="http://example.com/js/123.456789.chunk.js">
```

**Comment by alexander-akait** [comment]
Do you want to send a fix, it should be easy https://github.com/webpack/webpack/blob/main/lib/web/JsonpChunkLoadingRuntimeModule.js#L253

**Comment by snitin315** [comment]
@alexander-akait yes, can you assign this issue to me? Otherwise, I'll forget.

**Comment by alexander-akait** [comment]
Done :+1:

**Comment by snitin315** [comment]
@alexander-akait We need this only when `output.module: true` right? Also, how do I add a test case for this?

**Comment by alexander-akait** [comment]
@snitin315 Yeah, I think we already have test case for this, can you check?

**Comment by snitin315** [comment]
@alexander-akait We have for CJS https://github.com/webpack/webpack/blob/3c0af27ec6d5dc0ad33243a018068e5f1ab559cd/test/configCases/web/prefetch-preload/index.js#L30-L40

**Comment by alexander-akait** [comment]
@alexander-akait Yeah, just copy this test and set `experiments.outputModule` and check that we have right attributes
